src/librustc_mir/dataflow/generic.rs

```diff
@@ -0,0 +1,496 @@
+#![allow(unused)]
+
+use std::cmp::Ordering;
+use std::ops;
+
+use rustc::mir::{self, traversal, BasicBlock, Location};
+use rustc_data_structures::bit_set::BitSet;
+use rustc_data_structures::indexed_vec::{Idx, IndexVec};
+use rustc_data_structures::work_queue::WorkQueue;
+
+use crate::dataflow::BottomValue;
+
+/// A specific kind of dataflow analysis.
+///
+/// To run a dataflow analysis, one must set the initial state of the `START_BLOCK` via
+/// `initialize_start_block` and define a transfer function for each statement or terminator via
+/// the various `effect` methods. The entry set for all other basic blocks is initialized to
+/// `Self::BOTTOM_VALUE`. The dataflow `Engine` then iteratively updates the various entry sets for
+/// each block with the cumulative effects of the transfer functions of all preceding blocks.
+///
+/// You should use an `Engine` to actually run an analysis, and a `ResultsCursor` to inspect the
+/// results of that analysis like so:
+///
+/// ```ignore(cross-crate-imports)
+/// fn do_my_analysis(body: &mir::Body<'tcx>, dead_unwinds: &BitSet<BasicBlock>) {
+///     let analysis = MyAnalysis::new();
+///     let results = Engine::new(body, dead_unwinds, analysis).iterate_to_fixpoint();
+///     let mut cursor = dataflow::ResultsCursor::new(body, results);
+///
+///     for statement_index in body.block_data[START_BLOCK].statements.iter() {
+///         cursor.seek_after(Location { block: START_BLOCK, statement_index });
+///         let state = cursor.get();
+///         println!("{:?}", state);
+///     }
+/// }
+/// ```
+pub trait Analysis<'tcx>: BottomValue {
+    /// The index type used to access the dataflow state.
+    type Idx: Idx;
+
+    /// A name describing the dataflow analysis being implemented.
+    ///
+    /// The name should be suitable as part of a filename, so avoid whitespace, slashes or periods
+    /// and try to keep it short.
+    fn name() -> &'static str;
+
+    /// The size of each bitvector allocated for each block.
+    fn bits_per_block(&self, body: &mir::Body<'tcx>) -> usize;
```

Comment:
why does this depend on the body?

Reply:
It's not strictly necessary for most analyses since they also keep a reference to the `mir::Body`, but it hints to the user that the domain of the analysis depends on the body being analyzed.

One difference between `Analysis` and `BitDenotation` is that the various parts of the `mir::Body` (statements and terminators) are passed by reference to the `Analysis`. The old code required every analysis to store a reference to the body and do the lookup themselves. Simple analyses (like `HaveBeenBorrowedLocals`) wouldn't need to store a reference to the body as a field, they can operate on the same body that was passed to `Engine`. 

More complex analyses, like those that hold `MoveData` to deal with initialized places, have to do some computation ahead-of-time to figure out how big the domain is, and that computation will likely be shared between multiple analyses. They will ignore the `body` parameter here.